File: bindings/python/aster/interceptors/circuit_breaker.py
"""Circuit breaker interceptor."""

from __future__ import annotations

import time

from aster.interceptors.base import CallContext, Interceptor
from aster.status import RpcError, StatusCode


class CircuitBreakerInterceptor(Interceptor):
    """Simple CLOSED → OPEN → HALF_OPEN circuit breaker."""

    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"

    def __init__(
        self,
        *,
        failure_threshold: int = 3,
        recovery_timeout: float = 5.0,
        half_open_max_calls: int = 1,
    ) -> None:
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max_calls = half_open_max_calls
        self.state = self.CLOSED
        self.failure_count = 0
        self.opened_at = 0.0
        self.half_open_calls = 0

    def before_call(self, ctx: CallContext) -> None:
        now = time.monotonic()
        if self.state == self.OPEN:
            if now - self.opened_at >= self.recovery_timeout:
                self.state = self.HALF_OPEN
                self.half_open_calls = 0
            else:
                raise RpcError(StatusCode.UNAVAILABLE, "circuit breaker is open")

        if self.state == self.HALF_OPEN:
            if self.half_open_calls >= self.half_open_max_calls:
                raise RpcError(StatusCode.UNAVAILABLE, "circuit breaker is half-open")
            self.half_open_calls += 1

    def record_success(self) -> None:
        self.failure_count = 0
        self.half_open_calls = 0
        self.state = self.CLOSED

    def record_failure(self, error: RpcError) -> None:
        if error.code not in {StatusCode.UNAVAILABLE, StatusCode.INTERNAL, StatusCode.UNKNOWN}:
            return
        if self.state == self.HALF_OPEN:
            self.state = self.OPEN
            self.opened_at = time.monotonic()
            self.half_open_calls = 0
            return
        self.failure_count += 1
        if self.failure_count >= self.failure_threshold:
            self.state = self.OPEN
            self.opened_at = time.monotonic()

File: bindings/typescript/packages/aster/src/interceptors/circuit-breaker.ts
/**
 * Circuit breaker interceptor — stops sending requests to failing services.
 *
 * States: CLOSED (normal) -> OPEN (failing) -> HALF_OPEN (probe) -> CLOSED
 */

import type { Interceptor } from './base.js';
import { CallContext } from './base.js';
import { RpcError, StatusCode } from '../status.js';

type State = 'closed' | 'open' | 'half_open';

export interface CircuitBreakerOptions {
  failureThreshold?: number;
  resetTimeoutMs?: number;
  halfOpenMaxCalls?: number;
}

export class CircuitBreakerInterceptor implements Interceptor {
  private state: State = 'closed';
  private failures = 0;
  private lastFailure = 0;
  private halfOpenCalls = 0;

  private readonly failureThreshold: number;
  private readonly resetTimeoutMs: number;
  private readonly halfOpenMaxCalls: number;

  constructor(opts: CircuitBreakerOptions = {}) {
    this.failureThreshold = opts.failureThreshold ?? 5;
    this.resetTimeoutMs = opts.resetTimeoutMs ?? 30_000;
    this.halfOpenMaxCalls = opts.halfOpenMaxCalls ?? 1;
  }

  async onRequest(_ctx: CallContext, request: unknown): Promise<unknown> {
    if (this.state === 'open') {
      if (Date.now() - this.lastFailure > this.resetTimeoutMs) {
        this.state = 'half_open';
        this.halfOpenCalls = 0;
      } else {
        throw new RpcError(StatusCode.UNAVAILABLE, 'circuit breaker is open');
      }
    }

    if (this.state === 'half_open' && this.halfOpenCalls >= this.halfOpenMaxCalls) {
      throw new RpcError(StatusCode.UNAVAILABLE, 'circuit breaker half-open limit reached');
    }

    if (this.state === 'half_open') this.halfOpenCalls++;
    return request;
  }

  async onResponse(_ctx: CallContext, response: unknown): Promise<unknown> {
    if (this.state === 'half_open') {
      this.state = 'closed';
      this.failures = 0;
    }
    return response;
  }

  async onError(_ctx: CallContext, error: RpcError): Promise<RpcError> {
    this.failures++;
    this.lastFailure = Date.now();
    if (this.failures >= this.failureThreshold) {
      this.state = 'open';
    }
    return error;
  }

  /** Current circuit state. */
  get currentState(): State {
    return this.state;
  }

  /**
   * Pre-call gate check — throws if circuit is open.
   * Alias for the logic in onRequest(), callable without a full call context.
   */
  beforeCall(): void {
    if (this.state === 'open') {
      if (Date.now() - this.lastFailure > this.resetTimeoutMs) {
        this.state = 'half_open';
        this.halfOpenCalls = 0;
      } else {
        throw new RpcError(StatusCode.UNAVAILABLE, 'circuit breaker is open');
      }
    }
    if (this.state === 'half_open' && this.halfOpenCalls >= this.halfOpenMaxCalls) {
      throw new RpcError(StatusCode.UNAVAILABLE, 'circuit breaker half-open limit reached');
    }
    if (this.state === 'half_open') this.halfOpenCalls++;
  }

  /** Record a successful call — resets the failure count. */
  recordSuccess(): void {
    if (this.state === 'half_open') this.state = 'closed';
    this.failures = 0;
  }

  /** Record a failed call — may open the circuit. */
  recordFailure(): void {
    this.failures++;
    this.lastFailure = Date.now();
    if (this.failures >= this.failureThreshold) {
      this.state = 'open';
    }
  }
}
